include/winsparkle.h

```diff
@@ -242,6 +242,38 @@ WIN_SPARKLE_API void __cdecl win_sparkle_set_app_build_version(const wchar_t *bu
  */
 WIN_SPARKLE_API void __cdecl win_sparkle_set_registry_path(const char *path);
 
+/// Type used to override WinSparkle's config read function
+typedef int(__cdecl *win_sparkle_config_read_t)(void *custom_data, const char *name, wchar_t *buf, size_t len);
+/// Type used to override WinSparkle's config write function
+typedef int(__cdecl *win_sparkle_config_write_t)(void *custom_data, const char *name, const wchar_t *value);
+/// Type used to override WinSparkle's config delete function
+typedef int(__cdecl *win_sparkle_config_delete_t)(void *custom_data, const char *name);
+
+/**
+    Override WinSparkle's configuration read, write and delete functions.
+
+    By default, WinSparkle will read, write and delete configuration values by
+    interacting directly with Windows Registry.
+    If you want to manage configuration by yourself or you don't want let WinSparkle 
+    write settings directly to the Windows Registry, you can provide your own functions 
+    to read, write and delete configuration.
+
+    These functions needs to return TRUE on success, FALSE on failure.
+    If you passed NULL as a configuration action (read, write or delete)'s function pointer, 
+    WinSparkle will use the default function for that action.
+
+    @param custom_data  arbitrary data which will be passed to your config manipulating 
+                        functions, WinSparkle will not read or alter it.
+
+    @note There's no guarantee about the thread from which these functions are called,
```

Comment:
The calling code *does* provide some guarantees though: namely that all calls to the config functions are serialized (by means of `g_csConfigValues`) — i.e. you don't have to worry about thread-safely in the implementation most of the time, i.e. unless you use thread local storage for something.

Reply:
I was mainly concerning about the data race between WinSparkle's config functions and application's config functions.

For example if application stores all its configuration (including WinSparkle's settings) inside a single json file, then if application's main thread and WinSparkle's thread write contents to the file at same time, there will be a data race.